### PR TITLE
Add support to report Locality proximity information as part of Load assignment

### PR DESCRIPTION
**Pull request by vishalpowar**
Signed-off-by: Vishal Powar <vishalpowar@google.com>

Description: Adding API support to report Locality proximity information as part of Locality Assignment. This does not need to be documented yet, and should be documented as part of LB algorithm implementation which will use this information.

Risk Level: LOW
Testing: None
Docs Changes: None
Release Notes: None
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by zuercher** [comment]
Is the load balancer part of this sufficiently complicated that it needs to be in a separate review? It's fine if it is, but we should tag this field with `// [#not-implemented-hide:]` until there's an actual implementation.

**Comment by vishalpowar** [comment]
Will make the change to add [#not-implemented-hide:]

The LB algorithm needs a bit more discussion. The proximity information is at Locality granularity (proximity for endpoints within the locality does not really add a lot of value).

The current LBpolicy configuration has algorithms which do not distinguish between Locality and Endpoint.
With proximity we need a way to express LBpolicy at Locality and LBpolicy at Endpoint.

File: api/envoy/api/v2/endpoint/endpoint.proto [hunk 1]
```diff
@@ -132,4 +132,12 @@ message LocalityLbEndpoints {
...
   // Priorities should range from 0 (highest) to N (lowest) without skipping.
   uint32 priority = 5 [(validate.rules).uint32 = {lte: 128}];
+
+  // Optional: Per locality proximity value which indicates how close this
+  // locality is from the source locality. This value only provides ordering
+  // information (lower the value, closer it is to the source locality).
+  // This will be consumed by Load balancing schemes that need proximity order
+  // to determine where to route the requests.
+  // [#not-implemented-hide:]
+  google.protobuf.UInt32Value proximity_weight = 6;
```
Comment (htuch) [inline api/envoy/api/v2/endpoint/endpoint.proto]: Not wanting to bikeshed, but wondering if we want this to be [0.0, 1.0] weight or an integer? Whatever is most useful is good with me.

Comment (rshriram) [reply api/envoy/api/v2/endpoint/endpoint.proto]: nit: can we call this just proximity or locality_proximity instead of proximity weight?

**Comment by snowp** [comment]
Yeah I'm also interested in seeing how this will play with the existing load balancing algorithm. As @rshriram says we use priorities to model proximity because it provides exactly the "always hit the closest hosts, fail over if unavailable" semantics our legacy LB system used. 

It's not immediately clear to me how this would be used (ie will it have failover semantics? host selection weighting? both? is this just another way of specifying priorities?), but conceptually it seems like a reasonable thing to encode in the CLA.

File: api/envoy/api/v2/endpoint/endpoint.proto (hunk 1 above)
Comment (vishalpowar) [reply api/envoy/api/v2/endpoint/endpoint.proto]: This is semantically different than weight/pick-percent, integer values would better represent the absolute order [e.g. index of an array] we want to convey for these locality.

Changed to proximity

File: api/envoy/api/v2/endpoint/endpoint.proto [hunk 2]
```diff
@@ -132,4 +132,12 @@ message LocalityLbEndpoints {
   //
   // Priorities should range from 0 (highest) to N (lowest) without skipping.
   uint32 priority = 5 [(validate.rules).uint32 = {lte: 128}];
+
+  // Optional: Per locality proximity value which indicates how close this
+  // locality is from the source locality. This value only provides ordering
+  // information (lower the value, closer it is to the source locality).
+  // This will be consumed by Load balancing schemes that need proximity order
```
Comment (htuch) [inline api/envoy/api/v2/endpoint/endpoint.proto]: Nit: s/Load/load/

File: api/envoy/api/v2/endpoint/endpoint.proto [hunk 3]
```diff
@@ -132,4 +132,12 @@ message LocalityLbEndpoints {
   //
   // Priorities should range from 0 (highest) to N (lowest) without skipping.
   uint32 priority = 5 [(validate.rules).uint32 = {lte: 128}];
+
+  // Optional: Per locality proximity value which indicates how close this
```
Comment (zuercher) [inline api/envoy/api/v2/endpoint/endpoint.proto]: Nit: I think most people will assume smaller values mean closer proximity, but perhaps it should be explicitly stated?

Comment (vishalpowar) [reply api/envoy/api/v2/endpoint/endpoint.proto]: does the comment "(lower the value, closer it is to the source locality)" convey this information ?

File: api/envoy/api/v2/endpoint/endpoint.proto (hunk 2 above)
Comment (vishalpowar) [reply api/envoy/api/v2/endpoint/endpoint.proto]: Done

File: api/envoy/api/v2/endpoint/endpoint.proto (hunk 3 above)
Comment (zuercher) [reply api/envoy/api/v2/endpoint/endpoint.proto]: Oh, I'm sorry. It does say exactly what I wanted.

**Pull request merged**